feat: kubernetes overrides for the entrypoint and cmd

atchernych opened the pull request:
#### Overview:

Add the ability for a user to overwrite the entrypoint and PVCs in the sdk.
DEP-157

#### Details:

It is the first iteration of the feature.
A more generic approach will be developed.

#### Where should the reviewer start?

Still need a go test and type enforcement Model class for python

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added support for specifying a single main container with detailed Kubernetes configuration in component and graph deployments.
  - Users can now override container commands and arguments via Kubernetes-specific configuration fields.
  - Extended CRDs to include a comprehensive `mainContainer` specification for more granular pod customization.
- **Bug Fixes**
  - Corrected environment variable export syntax in deployment documentation.
- **Documentation**
  - Improved deployment instructions and added port forwarding guidance in the Hello World example.
  - Provided commented examples for Kubernetes container overrides in sample code.
- **Chores**
  - Updated YAML parsing dependency for improved compatibility and maintenance.
  - Updated attributions and dependency documentation to reflect new YAML library usage.
- **Style**
  - Fixed trailing blank lines in CRD YAML files for consistency.
- **Enhancements**
  - Added Docker socket mounting option for workspace support in container runtime.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

atchernych committed:
feat: rebase kubernetes overwrites

Overview:
Add the ability for a user to overwrite the entrypoint
and PVCs in the sdk.
DEP-157

Details:
It is the first iteration of the feature.
A more generic approach will be developed.

atchernych committed:
Changes per Julien

atchernych committed:
remove claim related stuff for now

atchernych committed:
cleanup

atchernych committed:
Dead-end changes

julienmancuso commented on deploy/cloud/operator/config/rbac/role.yaml:
```diff
@@ -1,18 +1,3 @@
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
```
we need this back

julienmancuso commented on deploy/cloud/operator/api/dynamo/common/common.go:
```diff
@@ -62,4 +62,8 @@ type ExtraPodSpec struct {
 	ServiceAccountName        string                            `json:"serviceAccountName,omitempty"`
 	PriorityClassName         string                            `json:"priorityClassName,omitempty"`
+
+	Volumes      []corev1.Volume                `json:"volumes,omitempty"`
```
I don't think these 2 are needed, are they ?

we just need to be able to take all these PVCs from [here](https://github.com/ai-dynamo/dynamo/pull/1396/files#diff-a943af46b6f5d3b0cf52240f727e566527b65e304e0df17046d7a04f28057caeR68) and mount them on the pod, right ?

julienmancuso commented on deploy/cloud/operator/api/v1alpha1/dynamocomponentdeployment_types.go:
```diff
@@ -79,6 +79,9 @@ type DynamoComponentDeploymentSharedSpec struct {
 	Replicas       *int32        `json:"replicas,omitempty"`
+
+	// +optional
+	Entrypoint *string `json:"entrypoint,omitempty"`
```
this should be removed and instead, 
in the `ExtraPodSpec` we should add a 
```
MainContainer                *corev1.Container `json:"mainContainer,omitempty"`
```
which would contain override for the main container.
For now it would be ok to only take care of overriding `Command` and `Args`

atchernych committed:
Refactored the python part after rebase

atchernych committed:
Move overrides into the extraPodSpecMainContainer

atchernych committed:
Cleanup after moving the overrides.

atchernych committed:
rerun make manifests

atchernych committed:
correct the manifest generation

atchernych committed:
fixed a bug with args

atchernych committed:
bring back licenses

atchernych commented:
/ok to test d7ce1cf1a0df17b1b52caf6d20c5822563e5cc0b

julienmancuso commented on deploy/cloud/operator/config/crd/bases/nvidia.com_dynamocomponentdeployments.yaml:
```diff
@@ -1606,6 +1606,684 @@ spec:
                       - name
                       type: object
                     type: array
+                  mainContainer:
```
you also need to copy all the CRDs in this directory to the `deploy/cloud/helm/crds/template`

and in the `nemo-dev-green` cluster (where the e2e tests are running) you need to manually update them by doing
`kubectl apply -f [FILE]`

atchernych committed:
Copy crds to deploy/cloud/helm/crds/template

atchernych committed:
Brign back  helm.sh/resource-policy: keep

atchernych committed:
add comments

atchernych committed:
add DynamoGraphDeploymentController changes

atchernych committed:
Fix the parsing of the ExtraPodSpec

atchernych committed:
fix up the overrides

julienmancuso commented on deploy/cloud/operator/internal/dynamo/graph.go:
```diff
@@ -360,6 +362,38 @@ func GenerateDynamoComponentsDeployments(ctx context.Context, parentDynamoGraphD
 		}
+
+		// Override command and args if provided.
+		if service.Config.ExtraPodSpec != nil && service.Config.ExtraPodSpec.MainContainer != nil {
```
please update the unit test in the associated `func TestGenerateDynamoComponentsDeployments(t *testing.T) {` to check the code in this if

julienmancuso replied:
actually this code needs to happen before mergo.Merge is called so that we can override configuration from `dynamo.yaml` with CRD content.

atchernych committed:
Move the ExtraPodSpec Overrides

atchernych committed:
proper naming for sharedSpec

julienmancuso commented on deploy/cloud/operator/internal/dynamo/graph.go:
```diff
@@ -368,13 +410,6 @@ func GenerateDynamoComponentsDeployments(ctx context.Context, parentDynamoGraphD
 			deployment.Spec.Autoscaling.MaxReplicas = service.Config.Autoscaling.MaxReplicas
 		}
-		// override the component config with the component config that is in the parent deployment
-		if configOverride, ok := parentDynamoGraphDeployment.Spec.Services[service.Name]; ok {
```
this need to stay here to be able to override the autoscaling part above.
also that doesn't sound right. the goal of the mergo step is to be able to override anything which is in config with whatever is set in the CRD. so in the CRD if mainContainer is not set, it shouldn't override what you put in the config file. I'm Trying to debug this issue

atchernych replied:
will do

atchernych commented on deploy/cloud/operator/internal/controller/dynamocomponentdeployment_controller.go:
```diff
@@ -1560,7 +1561,7 @@ func (r *DynamoComponentDeploymentReconciler) generatePodTemplateSpec(ctx contex
 		Image:          imageName,
 		Command:        []string{"sh", "-c"},
-		Args:           []string{strings.Join(args, " ")},
+		Args:           args,
```
will revert

atchernych commented on deploy/cloud/operator/internal/controller/dynamographdeployment_controller.go:
```diff
@@ -125,6 +126,19 @@ func (r *DynamoGraphDeploymentReconciler) Reconcile(ctx context.Context, req ctr
 
+	for _, service := range dynamoGraphConfig.Services {
+		if service.Config.ExtraPodSpec != nil {
+			extraPodSpecBytes, err := json.MarshalIndent(service.Config.ExtraPodSpec, "", "  ")
```
comments for debug. will remove.

atchernych committed:
move the override

override the component config with the component config
that is in the parent deployment should be after autoscaling.

atchernych committed:
revert default args

atchernych committed:
change the parsing library for yaml case

atchernych committed:
replace parsing lib and remove comments

atchernych committed:
Cleanup.

atchernych committed:
remove logger

atchernych committed:
Use BaseModel for KubernetesOverrides.

atchernych committed:
Add a go test for overrides.

atchernych committed:
refactor GenerateDynamoComponentsDeployments

julienmancuso commented on deploy/cloud/operator/internal/dynamo/graph.go:
```diff
@@ -527,3 +541,28 @@ func mergeEnvs(common, specific []corev1.EnvVar) []corev1.EnvVar {
+	}
+
+	if dst == nil {
+		panic("overrideExtraPodSpecMainContainer: dst must not be nil")
```
never panic, you should return an error instead.
if `deployment.Spec.DynamoComponentDeploymentSharedSpec.ExtraPodSpec` is nil, we should initialize it instead in this case

atchernych committed:
fix the help func

julienmancuso commented on deploy/cloud/operator/internal/dynamo/graph.go:
```diff
@@ -368,6 +370,19 @@ func GenerateDynamoComponentsDeployments(ctx context.Context, parentDynamoGraphD
+				deployment.Spec.DynamoComponentDeploymentSharedSpec.ExtraPodSpec = new(common.ExtraPodSpec)
+			}
+
+			err := overrideExtraPodSpecMainContainer(service.Config.ExtraPodSpec, &deployment.Spec.DynamoComponentDeploymentSharedSpec.ExtraPodSpec)
```
actually , we can remove your function and just use :
```
err := mergo.Merge(deployment.Spec.DynamoComponentDeploymentSharedSpec.ExtraPodSpec, service.Config.ExtraPodSpec, mergo.WithOverride)
if err != nil {
  return nil, err
}
```

atchernych replied:
I was a bit paranoid about the fact that we do not yet provide the overrides for all of the fields but we could control this upstream. Will Fix.

atchernych committed:
Streamline the ExtraPodEverrides

Change type for overriodes in python

atchernych committed:
validations on KubernetesOverrides

atchernych committed:
Simplify GenerateDynamoComponentsDeployments

fix build.py

atchernych committed:
Merge branch 'main' into dep-157-extend-decor

atchernych commented:
/ok to test 7a2b22792ae85804abb9f7748f2463e880f71156

atchernych committed:
Update run.sh

revert changes to run.sh

Signed-off-by: atchernych <atchernych@nvidia.com>

atchernych committed:
Merge branch 'main' into dep-157-extend-decor

atchernych committed:
Merge branch 'main' into dep-157-extend-decor

biswapanda commented on docs/examples/hello_world.md:
````diff
@@ -95,12 +95,19 @@ This example can be deployed to a Kubernetes cluster using [Dynamo Cloud](../../
 
 ```bash
-# Set your project root directory
+Make sure your dynamo cloud deploy.sh script from the prior step finished successfully and setup port forwaring in another window
````
I agree with the intent to mark the cloud deployment as pre-requisite. we already have a `### Prerequisites` section.  should we move these to relevant section?

* port forwarding
a.  Should we move the port forwarding instruction back to the pre-requisite section? 
can we provide a deterministic command, something like: 
```
kubectl port-forward svc/$DEPLOYMENT_NAME-frontend 3000:3000  -n ${KUBE_NS}
```

b.  typo with `forwaring` 
c. if there is an ingress user might not need to need to port-forward.


```
kubectl port-forward svc/...-dynamo-api-store <local-port>:80 -n $NAMESPACE
```

biswapanda replied:
to be fixed in follow up MR

biswapanda commented on deploy/sdk/src/dynamo/sdk/cli/build.py:
```diff
@@ -110,6 +112,7 @@ class ServiceConfig(BaseModel):
     dynamo: DynamoConfig = Field(default_factory=DynamoConfig)
     http_exposed: bool = False
     api_endpoints: t.List[str] = Field(default_factory=list)
+    kubernetes_overrides: KubernetesOverrides | None = None
```
is this `overrides` k8s specific? 
I think this is generic specification and it can be used in docker-compose / slurm etc. as well..

biswapanda reviewed (approved):
lgtm
- few comments doc/UX related comments

nv-anants reviewed (approved):
thanks for updating the attributions

atchernych committed:
Merge branch 'main' into dep-157-extend-decor
